# Dataset Cleaning
To reinforce your understanding of these concepts we are once again looking at the **Rain in Australia (weatherAUS)** dataset as to create a model that predicts rainfall given meteorological features.

## Imports
Below are the associated imports for the project. Don't be afraid to add more if to improve the quality of your model.

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from typing import List, Dict, Tuple
import numpy as np

## Dataset
Lets have a look into what features and aspects are found within our dataset. A method we've used already is `head` however another useful one to keep note of is `info` which gives information on the amount of `NaN` values. Where a nan value is just any data that isn't available.

For a more complex project it is advised to spend more time analysing the data as you may be able to extract important information useful for a model. For the sake of brevity we'll focus on filling and encoding data.

In [2]:
dataset = pd.read_csv("./data/weatherAUS.csv")
dataset.head()

,Date,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,WindDir9am,...,Humidity9am,Humidity3pm,Pressure9am,Pressure3pm,Cloud9am,Cloud3pm,Temp9am,Temp3pm,RainToday,RainTomorrow
0,2008-12-01,Albury,13.4,22.9,0.6,NaN,NaN,W,44.0,W,...,71.0,22.0,1007.7,1007.1,8.0,NaN,16.9,21.8,No,No
1,2008-12-02,Albury,7.4,25.1,0.0,NaN,NaN,WNW,44.0,NNW,...,44.0,25.0,1010.6,1007.8,NaN,NaN,17.2,24.3,No,No
2,2008-12-03,Albury,12.9,25.7,0.0,NaN,NaN,WSW,46.0,W,...,38.0,30.0,1007.6,1008.7,NaN,2.0,21.0,23.2,No,No
3,2008-12-04,Albury,9.2,28.0,0.0,NaN,NaN,NE,24.0,SE,...,45.0,16.0,1017.6,1012.8,NaN,NaN,18.1,26.5,No,No
4,2008-12-05,Albury,17.5,32.3,1.0,NaN,NaN,W,41.0,ENE,...,82.0,33.0,1010.8,1006.0,7.0,8.0,17.8,29.7,No,No


In [3]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

## **(Question 1)** Data-Preprocessing
### **(Part A)** Feature Extraction
Before we go about wrangling our data we first may want to choose features we believe are most important as to decrease the complexity of the training data. This will involves removing specific columns. Remember there are tradeoffs between both, with less features our model will train faster but at the cost of possibly missing important information. So when choosing be careful not to remove anything too vital. 

**Optionally remove columns to simplify the dataset**, make sure not to go overboard with this.

<details>
  <summary>Hint</summary>
    The function <code>dataset.drop([...], axis=1, inplace=True)</code> may be useful.
</details>

In [4]:
# Remove columns if you want.
dataset.drop(["Date"], axis=1, inplace=True) # Lazy Here

### **(Part B)** Not Available Data
From running `info` and `head` above we may see many incomplete fields within our dataset. For example despite their being 145460 entries only 142199 have a rain today value. Our first order of business is correcting these mistakes. There are two main approaches to this:
- Removing unused data.
- Replacing unused data (possibly with a calculation such as mean or median).

Be aware of the downsides of both strategies, when removing data you decrease its size which can mean you decrease the model's accuracy. However by filling the data you may be adding bias and simplifying its complexity. 

**Preprocess the data as to ensure there are no `NaN` values**.

<details>
  <summary>Hint</summary>
    Some functions which may be useful depending on the approach used are:
    <ul>
        <li><code>dataset[column].fillna(...)</code></li>
        <li><code>dataset[column].dropna()</code></li>
    </ul>
</details>

In [5]:
# Place modifications to the dataset here...
numeric_columns = ['MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation', 'Sunshine', 'WindGustSpeed', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Temp9am', 'Temp3pm']
for col in numeric_columns:
    dataset[col].fillna(dataset[col].median())
dataset.dropna(inplace=True)

To verify that you've removed all `NaN` data ensure values are 0 by running the below code.

In [6]:
dataset.isna().sum() # Verify by counting na

Location         0
MinTemp          0
MaxTemp          0
Rainfall         0
Evaporation      0
Sunshine         0
WindGustDir      0
WindGustSpeed    0
WindDir9am       0
WindDir3pm       0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Cloud9am         0
Cloud3pm         0
Temp9am          0
Temp3pm          0
RainToday        0
RainTomorrow     0
dtype: int64

### **(Part $\beta$)** Dates
This part is optional depending on if you removed the dates. In order to make use of our dates better it is best we split them into individual arguments rather than keeping them as one value. This will be easier to train on and be more generalisable in the long term.

**Split the dates into Days, Months, Years and remove Dates**. Given you haven't removed dates already.

In [7]:
# Apply modifications to the dataset here

This will print nothing when you have no longer have a date field.

In [8]:
if "Date" in dataset.columns:
    print("Please remove date")

### **(Part C)** Encoding Arguments
Looking back at the `dataset.info()` we see that a lot of data we are reliant on has the datatype `object`rather than boolean/integer. This data is mostly string labels, such as compass directions. This will cause problems later on as to generate weights and biases we need to encode these as numeric values. This can be done through two main approaches:
- Label encoding, where every category is encoded as an integer. This can be done automatically or done by using a relevant number to the data, such as the amount of appearences. 
- On hot encoding, where each category is represented as $n$ Boolean columns, where $n$ is the amount of categories. Below is an example with `RainToday`.

| ID | RainToday |
| ---- | ----- |
| 0 | Yes |
| 1 | No |

| ID | Yes | No |
| --- | ---- | ----- |
| 0 | 1 | 0 |
| 1 | 0 | 1 |

**Encode object arguments as integer values**, keep in mind columns that you don't encode need to be dropped as objects aren't a valid in neural networks.

<details>
  <summary>Hint</summary>
    Some functions which may be useful depending on the approach used are:
    <ul>
        <li><code>pd.get_dummies</code></li>
        <li><code>dataset[...].map({})</code></li>
        <li><code>OneHotEncoder</code> from sklearn</li>
        <li><code>LabelEncoder</code> from sklearn</li>
    </ul>
</details>

In [9]:
# Encode Dataset...
for x in ("RainToday", "RainTomorrow"):
    dataset[x] = dataset[x].map({"Yes": 1, "No": 0})
try:
    encoded_cols = pd.get_dummies(dataset[['Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm']])
    dataset = pd.concat([dataset, encoded_cols], axis=1)
    dataset.drop(['Location', 'WindGustDir', 'WindDir9am', 'WindDir3pm'], axis=1, inplace=True)
except: pass
boolean_columns = dataset.select_dtypes(include=bool).columns
dataset[boolean_columns] = dataset[boolean_columns].astype(int)

Given that you have gone about doing the above task nothing should be printed.

In [10]:
non_encoded_cols = dataset.select_dtypes(include=['object']).columns
if len(non_encoded_cols) > 0:
    print(f"The following rows need to be encoded:\n{non_encoded_cols}")

## Pytorch Datasets
After ensuring our dataset has the appropriate preprocessing done we can now turn it into a format that can be used by Torch. This is done with the `dataset` abstract class provided by Torch. For now we won't go too indepth just know that the below code creates a class that is able to access our data and then puts it into a iterator that allows for it to be returned by calling the `next(data)` command.

In [11]:
class WeatherAUS(Dataset):
    def __init__(self, data: pd.DataFrame):
        """Intialise the dataset with key information"""
        # If you want to modularise your model nicely it is fairly common to put
        # your initial preprocessing here. 
        labels = dataset["RainTomorrow"] # Use Rain Tomorrow as output
        d = data.drop(["RainTomorrow"], axis=1) # Remove it from dataset
        self.data = torch.tensor(d.values, dtype=torch.float32) # Convert data into tensors
        labels = torch.tensor(labels.values, dtype=torch.float32)
        self.labels = labels.view(-1, 1) # Turn Tensor[64] to Tensor[64, 1]

    def __len__(self) -> int:
        """Get length of the data"""
        return len(self.data)

    def __getitem__(self, idx: int) -> Tuple[Tensor, Tensor]:
        """Return the respective data at the index"""
        return self.data[idx], self.labels[idx]

# Apply wrapper to the dataset
data = WeatherAUS(dataset)